nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/FlattenJson.java

```diff
@@ -0,0 +1,134 @@
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License") you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.nifi.processors.standard;
+
+import com.github.wnameless.json.flattener.FlattenMode;
+import com.github.wnameless.json.flattener.JsonFlattener;
+import org.apache.nifi.annotation.documentation.CapabilityDescription;
+import org.apache.nifi.annotation.documentation.Tags;
+import org.apache.nifi.components.PropertyDescriptor;
+import org.apache.nifi.components.ValidationResult;
+import org.apache.nifi.flowfile.FlowFile;
+import org.apache.nifi.processor.AbstractProcessor;
+import org.apache.nifi.processor.ProcessContext;
+import org.apache.nifi.processor.ProcessSession;
+import org.apache.nifi.processor.ProcessorInitializationContext;
+import org.apache.nifi.processor.Relationship;
+import org.apache.nifi.processor.exception.ProcessException;
+
+import java.io.ByteArrayOutputStream;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+
+@Tags({ "json", "flatten" })
+@CapabilityDescription(
+        "Provides the user with the ability to take a nested JSON document and flatten it into a simple key/value pair " +
+        "document. The keys are combined at each level with a user-defined separator that defaults to '.'"
+)
+public class FlattenJson extends AbstractProcessor {
+    static final Relationship REL_SUCCESS = new Relationship.Builder()
+            .description("Successfully flattened files go to this relationship.")
+            .name("success")
+            .build();
+    static final Relationship REL_FAILURE = new Relationship.Builder()
+            .description("Files that cannot be flattened go to this relationship.")
+            .name("failure")
+            .build();
+    static final Relationship REL_ORIGINAL = new Relationship.Builder()
+            .description("The original files are copied to this relationship; autoterminated by default.")
+            .autoTerminateDefault(true)
+            .name("original")
+            .build();
+
+    static final PropertyDescriptor SEPARATOR = new PropertyDescriptor.Builder()
+            .name("flatten-json-separator")
+            .displayName("Separator")
+            .defaultValue(".")
+            .description("The separator character used for joining keys. Must be a JSON-legal character.")
+            .addValidator((subject, input, context) -> {
```

Comment:
I really wish we had a SINGLE_CHAR_VALIDATOR in StandardValidators :) For this case though, perhaps the validator should check to see if the character is JSON-legal? Better to find out before the processor is running.

Reply:
This validator should return "valid" if expression language (EL) is present, since it would be more than one character to express (see StandardValidators for some code to borrow). Also a unit test for this (using EL) would illustrate it well.

Reply:
The first comment above was incorporated (thanks!) but the second one doesn't seem to be. If you use EL here it looks like it will be marked invalid?

Reply:
@mattyb149 Take a look at the change I posted and see if you think it resolves it. Thanks.

Reply:
Did that just go in? I didn't see it when I wrote that comment, sorry if I was looking at something out of date. This should be fine; other validators usually just check if EL is present and mark it as valid. In this case since EL must resolve to a single character, then the entire property value must either be a single character or an EL expression, so this works here. 

Just to add some extra info, other EL-supporting properties can be partly literal and partly EL, where the entire property is not a valid expression but if EL is present, it can resolve to a valid value.

Reply:
I pushed in a change not that long after your comment.